File: houserobbery.py
# Runs on Leetcode
    # Runtime - O(n) - where n is length of nums
    # Memory - O(m*n) - m is length of nums, n is 2 since it's either choose ro not choose

class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        dp = [[None for i in range(2)] for j in range(len(nums))]
        
        rows = len(dp)
        cols = len(dp[0])
        
        dp[0][0] = 0
        dp[0][1] = nums[0]
        
        for i in range(1,rows):
            for j in range(cols):
                if j == 0:
                    dp[i][j] = max(dp[i-1][j+1],dp[i-1][j])
                else:
                    dp[i][j] = dp[i-1][j-1] + nums[i]
        
        return max(dp[rows-1][0],dp[rows-1][1])

File: houseRobbery.java
// Problem2 (https://leetcode.com/problems/house-robber/)

// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach in three sentences only
/*
 * In this problem, dynamic programming is implemented where using the 2 variables named prev and curr. curr variable will be moved across the
 * array from left ro right. prev would be 0th index element at start and curr would be max of prev index element and element at curr index,
 * for all the elements later curr is max of prev, prev + element at that index will be taken and prev is moved to previous curr. The curr
 * variable is returned.
 */

class Solution {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n == 1) return nums[0];

        int prev = nums[0];
        int curr = Math.max(prev, nums[1]); 

        for (int i = 2;i<nums.length;i++) {
            int temp = curr;
            curr = Math.max(temp, prev + nums[i]);
            prev = temp;

        }
        return curr;
    }
}